sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/_ownership_manager.py

```diff
@@ -0,0 +1,135 @@
+        4. Calculate the largest allowed count of partitions that an owner can own.
+        math.ceil(number of partitions / number of active owners). This should be equal or 1 greater than the average count
+        5. Adjust the number of partitions owned by this EventProcessor (owner)
+            a. if this EventProcessor owns more than largest allowed count, abandon one partition
+            b. if this EventProcessor owns less than average count, add one from the inactive or unclaimed partitions,
+                or steal one from another owner that has the largest number of ownership among all owners (EventProcessors)
+            c. Otherwise, no change to the ownership
+
+        The balancing algorithm adjust one partition at a time to gradually build the balanced ownership.
+        Ownership must be renewed to keep it active. So the returned result includes both existing ownership and
+        the newly adjusted ownership.
+        This method balances but doesn't claim ownership. The caller of this method tries to claim the result ownership
+        list. But it may not successfully claim all of them because of concurrency. Other EventProcessors may happen to
+        claim a partition at that time. Since balancing and claiming are run in infinite repeatedly,
+        it achieves balancing among all EventProcessors after some time of running.
+
+        :return: List[Dict[str, Any]], A list of ownership.
+        """
+        ownership_list = await self.partition_manager.list_ownership(self.eventhub_client.eh_name, self.consumer_group_name)
+        ownership_dict = dict((x["partition_id"], x) for x in ownership_list)  # put the list to dict for fast lookup
+
+        claimable_partition_ids = []  # partitions with inactive ownership and partitions that have never been claimed yet
+        active_ownership_self = []  # active ownership of this EventProcessor
+        active_ownership_count_group_by_owner = Counter()
+        for partition_id in self.all_parition_ids:
+            ownership = ownership_dict.get(partition_id)
+            if not ownership:  # no ownership found for this partition. So it is claimable
+                claimable_partition_ids.append(partition_id)
+            else:
+                last_modified_time = ownership["last_modified_time"]
+                owner_id = ownership["owner_id"]
+                now = time.time()
+                if now > self.ownership_timeout + last_modified_time:  # ownership timed out. So it is claimable
+                    claimable_partition_ids.append(partition_id)
+                else:  # the ownership is still active
+                    if owner_id == self.owner_id:  # partition is actively owned by this running EventProcessor
+                        active_ownership_self.append(ownership)
+                    active_ownership_count_group_by_owner[owner_id] = active_ownership_count_group_by_owner.get(owner_id, 0) + 1  # all active owners
```

Comment:
This is really odd usage of `collections.Counter`. It does it's own bookkeeping of counts (that is its purpose in life). You may just as well use a `set` if you are going to use it yourself...

Reply:
I used set before Counter. I changed to Counter because I want to use Counter.most_common( ) @ line 122. Counter receives an iterable in the constructor but it doesn't have a method to gradually receive individual values.